File: 0031-next-permutation/0031-next-permutation.java
import java.util.*;
class Solution {
    public void nextPermutation(int[] nums) {
        // go for longer prefix match by finding breakpoint
        int i = nums.length-2;
        while(i>=0){
            if(nums[i]<nums[i+1]) break;
            i--;
        }
        if(i==-1){
            i = 0;
            int j = nums.length-1;
            while(i<j){
                swap(nums, i, j);
                i++;
                j--;
            }
            return;
        }
        // now i is the element for which ceil is needed in its right subarray
        // find ceil of breakpoint 
        int j = nums.length-1;
        while(j>i){
            if(nums[j] > nums[i]) break;
            j--;
        }
        // now j is just bigger and we have to replace it
        swap(nums, i, j);
        // rearrange remaining in sorted order
        i = i+1;
        j = nums.length-1;
        while(i<j){
            swap(nums, i, j);
            i++;
            j--;
        }
    }
    private void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void swap(vector<int>& nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;  
    }
    void nextPermutation(vector<int>& nums) {
        int len = nums.size();
        int ind = -1;
        for(int i=len-2; i>=0; i--){
            if(nums[i] >= nums[i+1]) continue;
            else{
                ind = i;
                break;
            }
        }
        if(ind==-1){
            reverse(nums.begin(), nums.end());
            return;
        }
        int j = len-1;
        int i = ind;
        while(j>i){
            if(nums[j] > nums[i]) break;
            j--;
        }        
        swap(nums, i, j);
        //swap remaining.
        j = len-1;
        i++;
        while(i<j){
            swap(nums,i,j);
            i++;
            j--;
        }
    }
};


// 4 1 3 2  >>  4 3 1 2